lib/Backend/ServerThreadContext.cpp

```diff
@@ -4,19 +4,144 @@
 //-------------------------------------------------------------------------------------------------------
 
 #include "Backend.h"
+
 #if ENABLE_OOP_NATIVE_CODEGEN
 #include "JITServer/JITServer.h"
-#endif
+
+CriticalSection PageAllocatorPool::cs;
+PageAllocatorPool* PageAllocatorPool::Instance = nullptr;
+unsigned long long PageAllocatorPool::timerWating;
+
+PageAllocatorPool::PageAllocatorPool()
+    :pageAllocators(&NoThrowHeapAllocator::Instance),
+    activePageAllocatorCount(0)
+{
+    idleCleanupTimer = CreateWaitableTimerEx(NULL, L"JITServerIdle", 0/*auto reset*/, TIMER_ALL_ACCESS);
+}
+
+PageAllocatorPool::~PageAllocatorPool()
+{
+    RemoteAll();
+}
+
+void PageAllocatorPool::Initialize()
+{
+    Instance = HeapNewNoThrow(PageAllocatorPool);
+    if (Instance == nullptr)
+    {
+        Js::Throw::FatalInternalError();
+    }
+}
+void PageAllocatorPool::Shutdown()
+{
+    AutoCriticalSection autoCS(&cs);
+    HeapDelete(Instance);
+    Instance = nullptr;
+}
+void PageAllocatorPool::RemoteAll()
+{
+    AutoCriticalSection autoCS(&cs);
+    while (!pageAllocators.Empty())
+    {
+        HeapDelete(pageAllocators.Pop());
+    }
+}
+
+void PageAllocatorPool::IncreaseActiveCount()
+{
+    InterlockedIncrement(&activePageAllocatorCount);
+}
+void PageAllocatorPool::DecreaseActiveCount()
+{
+    unsigned long long currentActiveCount = InterlockedDecrement(&PageAllocatorPool::Instance->activePageAllocatorCount);
+    if (currentActiveCount == 0)
+    {
+        PageAllocatorPool::IdleCleanup();
+    }
+}
+
+PageAllocator* PageAllocatorPool::GetPageAllocator()
+{
+    AutoCriticalSection autoCS(&cs);
+    PageAllocator* pageAllocator = nullptr;
+    if (pageAllocators.Count() > 0)
+    {
+        // TODO: OOP JIT, select the page allocator with right count of free pages
+        // base on some heuristic
+        pageAllocator = this->pageAllocators.Pop();
+    }
+    else
+    {
+        pageAllocator = HeapNew(PageAllocator, nullptr, Js::Configuration::Global.flags, PageAllocatorType_BGJIT,
+            AutoSystemInfo::Data.IsLowMemoryProcess() ? PageAllocator::DefaultLowMaxFreePageCount : PageAllocator::DefaultMaxFreePageCount);
```

Comment:
>IsLowMemoryProcess [](start = 33, length = 18)

This value in the server might bedifferent from the client process.  We should get this from the client process instead.